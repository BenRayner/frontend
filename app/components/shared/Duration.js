// @flow

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { second } from 'metrick/duration';
import moment from 'moment';
import { getDurationString } from '../../lib/date';

type Props = {
  from?: (moment$Moment | string | number | Date | number[]),
  to?: (moment$Moment | string | number | Date | number[]),
  className?: string,
  tabularNumerals: boolean,
  format: getDurationString.formats,
  updateFrequency?: number
};

type State = {
  value: string
};

const Moment = moment().constructor;

class Duration extends React.PureComponent<Props, State> {
  static propTypes = {
    from: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.instanceOf(Date),
      PropTypes.instanceOf(Moment)
    ]),
    to: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.instanceOf(Date),
      PropTypes.instanceOf(Moment)
    ]),
    className: PropTypes.string,
    tabularNumerals: PropTypes.bool.isRequired,
    format: PropTypes.oneOf(getDurationString.formats),
    updateFrequency: PropTypes.number
  };

  static defaultProps = {
    updateFrequency: second.bind(1),
    tabularNumerals: true
  };

  _interval: IntervalID;

  state = {
    value: ''
  };

  updateTime() {
    const { from, to, format } = this.props;

    this.setState({
      value: getDurationString(from, to, format)
    });
  }

  componentDidMount() {
    this.maybeSetInterval(this.props.updateFrequency);
  }

  maybeSetInterval(updateFrequency) {
    if (typeof updateFrequency == 'number' && updateFrequency > 0) {
      this._interval = setInterval(() => this.updateTime(), updateFrequency);
    }
    this.updateTime();
  }

  maybeClearInterval() {
    if (this._interval) {
      clearInterval(this._interval);
    }
  }

  componentWillUnmount() {
    this.maybeClearInterval();
  }

  componentWillReceiveProps(nextProps) {
    const { from, to, format, updateFrequency } = nextProps;

    if (updateFrequency !== this.props.updateFrequency) {
      this.maybeClearInterval();
      this.maybeSetInterval(updateFrequency);
    }

    this.setState({
      value: getDurationString(from, to, format)
    });
  }

  render() {
    const { state: { value }, props: { className, tabularNumerals } } = this;
    const spanClassName = classNames(
      className,
      { 'tabular-numerals': tabularNumerals }
    );

    return (
      <span className={spanClassName}>
        {value}
      </span>
    );
  }
}

const exported = {};

getDurationString.formats.forEach((format) => {
  const componentName = format.charAt(0).toUpperCase() + format.slice(1);

  const component = (props) => <Duration {...props} format={format} />;
  component.displayName = `Duration.${componentName}`;

  exported[componentName] = component;
});

export default exported;
