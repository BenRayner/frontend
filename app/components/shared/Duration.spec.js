/* global jest, describe, it, expect, xdescribe */
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';

import Duration from './Duration';

jest.mock('../../lib/date', () => {
  const getDurationString = jest.fn(() => 'MOCKED-DURATION');
  getDurationString.formats = ['expected'];
  const getDuration = jest.fn(() => 42069);
  return { getDuration, getDurationString };
});

describe('Duration', () => {
  const componentList = Object.keys(Duration);
  const DurationComponent = Duration.Expected;

  it('provides stateless Components named after `getDurationString.formats`', () => {
    expect(componentList).toEqual(['Expected']);
  });

  it('calls through to `getDurationString`', () => {
    const getDuration = require('../../lib/date').getDuration;
    const getDurationString = require('../../lib/date').getDurationString;
    const from = new Date();
    const to = new Date();

    const component = ReactTestRenderer.create(
      <DurationComponent
        from={from}
        to={to}
        updateFrequency={0}
      />
    );

    const tree = component.toJSON();
    expect(getDuration).toHaveBeenCalledWith(from, to);
    expect(getDurationString).toHaveBeenCalledWith(42069, 'expected');
    expect(tree).toMatchSnapshot();
  });

  describe('tabularNumerals', () => {
    it('can be disabled', () => {
      const getDuration = require('../../lib/date').getDuration;
      const getDurationString = require('../../lib/date').getDurationString;
      const from = new Date();
      const to = new Date();

      const component = ReactTestRenderer.create(
        <DurationComponent
          tabularNumerals={false}
          from={from}
          to={to}
          updateFrequency={0}
        />
      );

      const tree = component.toJSON();
      expect(getDuration).toHaveBeenCalledWith(from, to);
      expect(getDurationString).toHaveBeenCalledWith(42069, 'expected');
      expect(tree).toMatchSnapshot();
    });
  });

  xdescribe('updateFrequency', () => {
    // TODO: Need to be able to instrument updating and interactions
    it('sets an interval if supplied with a frequency greater than zero', () => {
      const rendered = ReactTestRenderer.create(
        <DurationComponent
          from="2016-05-07T09:00:00.000Z"
          to="2016-05-07T09:00:00.000Z"
          updateFrequency={10}
        />
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });

    it('sets no interval if supplied with a frequency of zero', () => {
      const rendered = ReactTestRenderer.create(
        <DurationComponent
          from="2016-05-07T09:00:00.000Z"
          to="2016-05-07T09:00:00.000Z"
          updateFrequency={0}
        />
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });
});
