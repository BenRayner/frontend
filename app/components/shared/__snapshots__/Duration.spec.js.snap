exports[`Duration Duration.Full renders as expected 1`] = `
<span>
  0 seconds
</span>
`;

exports[`Duration Duration.Full renders as expected 2`] = `
<span>
  5 seconds
</span>
`;

exports[`Duration Duration.Full renders as expected 3`] = `
<span>
  15 minutes, 7 seconds
</span>
`;

exports[`Duration Duration.Full renders as expected 4`] = `
<span>
  1 hour, 45 minutes, 16 seconds
</span>
`;

exports[`Duration Duration.Full renders as expected 5`] = `
<span>
  4 hours, 59 minutes, 3 seconds
</span>
`;

exports[`Duration Duration.Full renders as expected 6`] = `
<span>
  1 day, 7 hours, 3 minutes
</span>
`;

exports[`Duration Duration.Full renders as expected 7`] = `
<span>
  1 week, 13 days, 16 hours
</span>
`;

exports[`Duration Duration.Full renders as expected 8`] = `
<span>
  9 weeks, 2 days, 20 hours
</span>
`;

exports[`Duration Duration.Micro renders as expected 1`] = `
<span>
  0s
</span>
`;

exports[`Duration Duration.Micro renders as expected 2`] = `
<span>
  5s
</span>
`;

exports[`Duration Duration.Micro renders as expected 3`] = `
<span>
  15m
</span>
`;

exports[`Duration Duration.Micro renders as expected 4`] = `
<span>
  2h
</span>
`;

exports[`Duration Duration.Micro renders as expected 5`] = `
<span>
  5h
</span>
`;

exports[`Duration Duration.Micro renders as expected 6`] = `
<span>
  1d
</span>
`;

exports[`Duration Duration.Micro renders as expected 7`] = `
<span>
  3w
</span>
`;

exports[`Duration Duration.Micro renders as expected 8`] = `
<span>
  9w
</span>
`;

exports[`Duration Duration.Short renders as expected 1`] = `
<span>
  0s
</span>
`;

exports[`Duration Duration.Short renders as expected 2`] = `
<span>
  5s
</span>
`;

exports[`Duration Duration.Short renders as expected 3`] = `
<span>
  15m 7s
</span>
`;

exports[`Duration Duration.Short renders as expected 4`] = `
<span>
  1h 45m
</span>
`;

exports[`Duration Duration.Short renders as expected 5`] = `
<span>
  4h 59m
</span>
`;

exports[`Duration Duration.Short renders as expected 6`] = `
<span>
  1d 7h
</span>
`;

exports[`Duration Duration.Short renders as expected 7`] = `
<span>
  1w 14d
</span>
`;

exports[`Duration Duration.Short renders as expected 8`] = `
<span>
  9w 3d
</span>
`;

exports[`Duration stateless Components are accessible 1`] = `
Array [
  "Full",
  "Short",
  "Micro",
]
`;
